File: Top interview 150/0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        # BFS를 위한 큐 초기화, 시작 노드 추가
        queue = deque([root])
        # 오른쪽에서 보이는 노드들을 저장할 결과 리스트
        res = []

        # 큐가 빌 때까지 반복 (모든 레벨을 순회)
        while queue:
            # 현재 레벨의 노드 개수
            level_size = len(queue)
            
            # 현재 레벨의 모든 노드를 처리
            for i in range(level_size):
                node = queue.popleft()  # 큐에서 노드 추출
                
                # 현재 레벨의 마지막 노드인 경우 (= 오른쪽에서 보이는 노드)
                if i == level_size - 1:
                    res.append(node.val)
                
                # 왼쪽 자식을 큐에 추가
                if node.left:
                    queue.append(node.left)
                # 오른쪽 자식을 큐에 추가
                if node.right:
                    queue.append(node.right)

        return res

File: LeetCode 75/0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<int> RightSideView(TreeNode root) {
        // 결과를 저장할 리스트 초기화
        var result = new List<int>();
        
        if (root == null) {
            return result;
        }
        
        // BFS를 위한 큐 초기화, 시작 노드 추가
        var queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        
        // 큐가 빌 때까지 반복 (모든 레벨을 순회)
        while (queue.Count > 0) {
            // 현재 레벨의 노드 개수
            int levelSize = queue.Count;
            
            // 현재 레벨의 모든 노드를 처리
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.Dequeue();
                
                // 현재 레벨의 마지막 노드인 경우 (= 오른쪽에서 보이는 노드)
                if (i == levelSize - 1) {
                    result.Add(node.val);
                }
                
                // 왼쪽 자식을 큐에 추가
                if (node.left != null) {
                    queue.Enqueue(node.left);
                }
                // 오른쪽 자식을 큐에 추가
                if (node.right != null) {
                    queue.Enqueue(node.right);
                }
            }
        }
        
        return result;
    }
}